mountpoint-s3/src/data_cache/multilevel_cache.rs

```diff
@@ -0,0 +1,323 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use futures::task::{Spawn, SpawnExt};
+use tracing::{debug, warn};
+
+use crate::object::ObjectId;
+
+use super::{BlockIndex, ChecksummedBytes, DataCache, DataCacheResult};
+
+pub struct MultilevelDataCache<DiskCache, ExpressCache, Runtime> {
+    disk_cache: Arc<DiskCache>,
+    express_cache: ExpressCache,
+    runtime: Runtime,
+}
+
+impl<DiskCache: DataCache, ExpressCache: DataCache, Runtime: Spawn>
+    MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    pub fn new(disk_cache: Arc<DiskCache>, express_cache: ExpressCache, runtime: Runtime) -> Self {
+        assert_eq!(disk_cache.block_size(), express_cache.block_size());
+        Self {
+            disk_cache,
+            express_cache,
+            runtime,
+        }
+    }
+}
+
+#[async_trait]
+impl<DiskCache: DataCache + Sync + Send + 'static, ExpressCache: DataCache + Sync, Runtime: Spawn + Sync> DataCache
+    for MultilevelDataCache<DiskCache, ExpressCache, Runtime>
+{
+    async fn get_block(
+        &self,
+        cache_key: &ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+        object_size: usize,
+    ) -> DataCacheResult<Option<ChecksummedBytes>> {
+        if let Some(data) = self
+            .disk_cache
+            .get_block(cache_key, block_idx, block_offset, object_size)
+            .await?
+        {
+            debug!(cache_key=?cache_key, block_idx=block_idx, "block served from the disk cache");
+            return DataCacheResult::Ok(Some(data));
+        }
+        if let Some(data) = self
+            .express_cache
+            .get_block(cache_key, block_idx, block_offset, object_size)
+            .await?
+        {
+            debug!(cache_key=?cache_key, block_idx=block_idx, "block served from the express cache");
+            let cache_key = cache_key.clone();
+            let cache = self.disk_cache.clone();
+            let data_cloned = data.clone();
+            self.runtime
+                .spawn(async move {
+                    if let Err(error) = cache
+                        .put_block(cache_key.clone(), block_idx, block_offset, data_cloned, object_size)
+                        .await
+                    {
+                        warn!(cache_key=?cache_key, block_idx, ?error, "failed to update cache");
+                    }
+                })
+                .unwrap();
+            return DataCacheResult::Ok(Some(data));
+        }
+        DataCacheResult::Ok(None)
+    }
+
+    async fn put_block(
+        &self,
+        cache_key: ObjectId,
+        block_idx: BlockIndex,
+        block_offset: u64,
+        bytes: ChecksummedBytes,
+        object_size: usize,
+    ) -> DataCacheResult<()> {
+        self.disk_cache
```

Comment:
Could these be done in parallel? Should they be?

Reply:
it's definitely an option with something like
```rust
    async fn put_block(
        &self,
        cache_key: ObjectId,
        block_idx: BlockIndex,
        block_offset: u64,
        bytes: ChecksummedBytes,
        object_size: usize,
    ) -> DataCacheResult<()> {
        let fut1= self.disk_cache
            .put_block(cache_key.clone(), block_idx, block_offset, bytes.clone(), object_size);
        let fut2= self.express_cache
            .put_block(cache_key, block_idx, block_offset, bytes, object_size);
        let (res1, res2) = join!(fut1, fut2);
        res1?;
        res2?;
        Ok(())
    }
```

it may result in express cache being populated quicker, making data available to other instances, but the difference will not be too big

Reply:
Decided to execute them sequentially. The reasoning is:
1. we want the local cache for this instance to be available as soon as possible as it provides the best performance benefit on the repeated access to a block;
2. we are ok with having a delay when populating the shared cache, as it'll be used by other instances of MP and they may have their local cache;
3. starting the S3 request concurrently with the disk write may result in a longer delay for the disk operation.